ArduPlane/system.cpp

```diff
@@ -246,15 +246,8 @@ bool Plane::set_mode(Mode &new_mode, const ModeReason reason)
         // we failed entering new mode, roll back to old
         previous_mode = &old_previous_mode;
         control_mode = &old_mode;
-
         control_mode_reason = previous_mode_reason;
 
-        // currently, only Q modes can fail enter(). This will likely change in the future and all modes
-        // should be changed to check dependencies and fail early before depending on changes in Mode::set_mode()
-        if (control_mode->is_vtol_mode()) {
-            // ignore result because if we fail we risk looping at the qautotune check above
-            control_mode->enter();
```

Comment:
This could be very, very bad.

We reset a whole slew of stuff in the top if `enter` - this call was supposed to undo a whole bunch of that.

It's a seriously crap pattern we currently have, but to remove the call here we'd have to undo all of the setting we do in the top of `Mode::enter` https://github.com/ardupilot/ardupilot/blob/master/ArduPlane/mode.cpp#L18

Reply:
We already do all of this in master, and we can't fix everything all the time